atomic_reactor/dirs.py

```diff
@@ -0,0 +1,227 @@
+            return dict(item.split("=", 1) for item in envs)
+
+    def dockerfile_with_parent_env(self, parent_inspect: ImageInspectionData) -> DockerfileParser:
+        """Get the parsed Dockerfile with parent information injected.
+
+        :param parent_inspect: a mapping containing the image inspection data,
+            which is used as parent image's inspection data to be injected into
+            the parsed Dockerfile.
+        :type parent_inspect: dict[str, any]
+        :return: the parsed Dockerfile
+        :rtype: DockerfileParser
+        """
+        envs = self._get_env_from_inspection(parent_inspect)
+        if envs is None:
+            logger.debug("Parent Environment not found, not applied to Dockerfile")
+        return DockerfileParser(str(self.dockerfile_path), parent_env=envs)
+
+
+class RootBuildDir(object):
+    """A directory containing all artifacts for building images.
+
+    This directory contains two kinds of directories, one is the container
+    repository cloned from the remote, and the other one is the
+    platform-specific copies. The following is an overview of a typical
+    directory structure:
+
+    /path/to/root_build_dir
+    |-- the original source dir
+    |-- aarch64/
+    |-- ppc64le/
+    |-- s390x/
+    +-- x86_64/
+    """
+
+    def __init__(self, path: Path, platforms: List[Platform]) -> None:
+        if not path.exists():
+            raise FileNotFoundError(f"Path {path} does not exist.")
+        self.path = path
+        if not platforms:
+            raise ValueError("At least one platform must be specified.")
+        self.platforms = platforms
+
+    def copy_sources(self, source: Source) -> None:
+        """Create platform-specific build directories from source.
+
+        :param source: the original source from where to copy the content into
+            every platform-specific directory.
+        :type source: Source
+        """
+        src_path = source.path
+        ignore = ignore_patterns(".git")
+        for platform in self.platforms:
+            copytree(src_path, self.path / platform.value, ignore=ignore)
+
+    @property
+    def has_sources(self) -> bool:
+        """Check if all platform-specific directories exist.
+
+        :return: True if all platform-specific directories exist. Otherwise,
+            False is returned.
+        :rtype: bool
+        """
+        for platform in self.platforms:
+            if not self.path.joinpath(platform.value).exists():
+                return False
+        return True
+
+    @property
+    def any_build_dir(self) -> BuildDir:
+        """Get a platform-specific build directory.
+
+        This is typically used by code that does not care about platform. The
+        returned directory is guaranteed to be the same one between the
+        multiple calls.
+
+        :return: a source directory.
+        :rtype: BuildDir
+        """
+        platform: Platform = sorted(self.platforms, key=lambda item: item.value)[0]
```

Comment:
Maybe sort the platforms aleady in `__init__`? IMO that would also make the `for_each()` and `for_all_copy()` methods slightly more predictable (though the order still shouldn't be relied on)

Reply:
> I think we lack information whether the order of passed platforms matters for the caller

In my opinion, we should never have code that relies on processing the platforms in a specific order

> For what cases of this predictability?

We will have at least two different ways to create the RootBuildDir - in the prebuild task, it will be done by the `check_and_set_platforms` plugin, in all other tasks after that, it will be done differently. If the order is up to the caller, then we would have to make sure every initialization passes the platforms in the same order, otherwise different tasks could process platforms in different order (which wouldn't matter that much, since IMO the order shouldn't matter, but it would still be nice to ensure consistency).